runtime/format.h

```diff
@@ -0,0 +1,108 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// FORMAT string processing
+
+#ifndef FORTRAN_RUNTIME_FORMAT_H_
+#define FORTRAN_RUNTIME_FORMAT_H_
+
+#include "terminator.h"
+#include "../lib/common/Fortran.h"
+#include <cinttypes>
+#include <optional>
+
+namespace Fortran::runtime {
+
+enum EditingFlags {
+  blankZero = 1,  // BLANK=ZERO or BZ edit
+  decimalComma = 2,  // DECIMAL=COMMA or DC edit
+  signPlus = 4,  // SIGN=PLUS or SP edit
+};
+
+struct MutableModes {
+  std::uint8_t editingFlags{0};  // BN, DP, SS
+  common::RoundingMode roundingMode{common::RoundingMode::TiesToEven};  // RN
+};
+
+// A single edit descriptor extracted from a FORMAT
+struct DataEdit {
+  char descriptor;  // capitalized: one of A, I, B, O, Z, F, E(N/S/X), D, G
+  char variation{'\0'};  // N, S, or X for EN, ES, EX
+  int width;  // the 'w' field
+  std::optional<int> digits;  // the 'm' or 'd' field
+  std::optional<int> expoDigits;  // 'Ee' field
+  MutableModes modes;
+};
+
+// Generates a sequence of DataEdits from a FORMAT statement or
+// default-CHARACTER string.  Driven by I/O item list processing.
+// Errors are fatal.  See clause 13.4 in Fortran 2018 for background.
+class FormatControl {
+public:
+  struct Context {
+    Terminator &terminator;
+    void (*handleCharacterLiteral)(const char *, std::size_t){nullptr};
+    void (*handleSlash)(){nullptr};
+    void (*handleAbsolutePosition)(int){nullptr};  // Tn
+    void (*handleRelativePosition)(int){nullptr};  // nX, TRn, TLn (negated)
+  };
+
+  FormatControl(Context &, const char *format, std::size_t formatLength,
+      const MutableModes &initialModes, int maxHeight = maxMaxHeight);
+
+  // Determines the max parenthesis nesting level by scanning and validating
+  // the FORMAT string.
+  static int GetMaxParenthesisNesting(
+      Terminator &, const char *format, std::size_t formatLength);
+
+  // For attempting to allocate in a user-supplied stack area
+  static std::size_t GetNeededSize(int maxHeight) {
+    return sizeof(FormatControl) -
+        sizeof(Iteration) * (maxMaxHeight - maxHeight);
+  }
+
+  // Extracts the next data edit descriptor, handling control edit descriptors
+  // along the way.
+  void GetNext(DataEdit &);
+
+  // Emit any remaining character literals after the last data item.
+  void FinishOutput();
+
+private:
+  static constexpr std::uint8_t maxMaxHeight{100};
+
+  struct Iteration {
+    static constexpr int unlimited{-1};
+    int start{0};  // offset in format_ of '(' or a repeated edit descriptor
+    int remaining{0};  // while >0, decrement and iterate
+  };
+
+  char PeekNext() { return offset_ < formatLength_ ? format_[offset_] : '\0'; }
```

Comment:
Does PeekNext need to account for intervening white space?

Reply:
No, that's a responsibility of its callers.

Reply:
Does the call in GetIntField do that?  At least a couple of cases have SkipBlanks predecessors that wouldn't be needed if PeekNext were able to skip.  (Assuming that's not problematic.)

Reply:
No, it doesn't.  Do you think that it should?  I should either completely ignore white space outside character literals and Hollerith, or just allow for it between descriptors and punctuation.  Not sure what the best policy would be.

Reply:
Since 1 2  34 5  is just as valid as 12345, even if numeric values aren't likely to appear that way very often, that does need to be accepted.  In that case, the code needs to either make sure all PeekNext calls have an associated SkipBlanks predecessor or some other way to skip blanks.  It may be more robust to have PeekNext do that itself, but of course that isn't required.  Another option would be for FormatValidator  to squeeze blanks (could even be done in situ), but that would complicate error message generation.  My suggestion is to pick a simple, but correct solution.  I was under the impression that we weren't too concerned about repeated format scanning, even for the case Jean noticed.  For that case, a star repetition count could be represented by an appropriate MAXINT.

Reply:
Sounds good to me, will do.  Thanks!